src/H5Dmpio.c

```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+
+    FUNC_ENTER_STATIC
+
+    HDassert(chunk_list || 0 == chunk_list_num_entries);
+    HDassert(io_info);
+    HDassert(idx_info);
+
+#ifdef H5Dmpio_DEBUG
+    H5D_MPIO_TRACE_ENTER(mpi_rank);
+    H5D_MPIO_TIME_START(mpi_rank, "Reallocation of chunk file space");
+#endif
+
+    /* Create derived datatypes for the chunk file space info needed */
+    if (H5D__mpio_get_chunk_fspace_info_types(&recv_type, &recv_type_derived, &send_type,
+                                              &send_type_derived) < 0)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL,
+                    "can't create derived datatypes for chunk file space info")
+
+    /* Gather the new chunk sizes to all ranks for a collective reallocation
+     * of the chunks in the file.
+     */
+    H5_CHECK_OVERFLOW(chunk_list_num_entries, size_t, int);
+    if (H5_mpio_gatherv_alloc_simple(chunk_list, (int)chunk_list_num_entries, send_type, recv_type, TRUE, 0,
+                                     io_info->comm, mpi_rank, mpi_size, &gathered_array,
+                                     &collective_num_entries) < 0)
+        HGOTO_ERROR(H5E_DATASET, H5E_CANTGATHER, FAIL, "can't gather chunk file space info to/from ranks")
+    collective_list = (H5D_filtered_collective_io_fspace_info_t *)gathered_array;
+
+    if (collective_num_entries) {
+        H5D_filtered_collective_io_fspace_info_t *coll_entry;
+        hbool_t                                   need_insert;
+
+        /* Collectively re-allocate the modified chunks (from each rank) in the file */
+        HDassert(idx_info->storage->idx_type != H5D_CHUNK_IDX_NONE);
+        for (i = 0; i < collective_num_entries; i++) {
+            coll_entry = &collective_list[i];
+
+            /*
+             * Once we find the section of chunks that belong to this rank,
+             * do the file space re-allocation for each chunk and keep the
+             * local chunk list updated alongside the allocations. This saves
+             * a little on post-processing and memory usage.
+             */
+            if (chunk_list_num_entries &&
+                coll_entry->chunk_current.offset == chunk_list[0].fspace_info.chunk_current.offset) {
+                H5D_filtered_collective_io_fspace_info_t *local_entry;
+                size_t                                    num_processed = 0;
+
+                do {
+                    local_entry = &chunk_list[num_processed].fspace_info;
+                    coll_entry  = &collective_list[i];
+
+                    if (H5D__chunk_file_alloc(idx_info, &coll_entry->chunk_current, &coll_entry->chunk_new,
+                                              &chunk_list[num_processed].index_info.need_insert, NULL) < 0)
+                        HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "unable to allocate chunk")
+                    *local_entry = *coll_entry;
+                } while ((++num_processed < chunk_list_num_entries) && (++i < collective_num_entries));
+
+                continue;
```

Comment:
Yes, the `continue` was more to just make this slightly odd section of code stand out, since it might be non-obvious what it's doing. If you'd prefer a simpler `else` for readability, I can change that.

Reply:
Made much cleaner in latest changes.